File: myurlshortner-consumer/myurlshortner-consumer/src/main/scala/com/acme/myurlshortner/consumer/application/service/ShortenedUrlUserEventsServiceImpl.scala
package com.acme.myurlshortner.consumer.application.service

import com.acme.myurlshortner.consumer.application.repo.ShortenedUrlUserAccessRepository
import com.acme.myurlshortner.consumer.domain.command.SaveShortenedUrlUserAccessCommand
import com.acme.myurlshortner.consumer.domain.entity.useragent._
import com.acme.myurlshortner.consumer.domain.entity.userevent.ShortenedUrlUserAccess
import com.acme.myurlshortner.consumer.domain.service.ShortenedUrlUserEventsService
import zio.ZIO
import zio.ZLayer

class ShortenedUrlUserEventsServiceImpl(
  val shortenedUrlUserAccessRepository: ShortenedUrlUserAccessRepository
) extends ShortenedUrlUserEventsService {

  override def saveUserAccess(command: SaveShortenedUrlUserAccessCommand): ZIO[Any, Throwable, Unit] =
    for {
      browser    <- ZIO.succeed(
                      command.userAgent match
                        case a if a.contains(MozillaFirefox.userAgentValue) => MozillaFirefox
                        case b if b.contains(Chrome.userAgentValue)         => Chrome
                        case c if c.contains(Safari.userAgentValue)         => Safari
                        case _                                              => OtherBrowser
                    )
      device     <- ZIO.succeed(
                      command.userAgent match
                        case a if a.contains(Mac.userAgentValue)     => Mac
                        case b if b.contains(iPad.userAgentValue)    => iPad
                        case c if c.contains(iPhone.userAgentValue)  => iPhone
                        case d if d.contains(Android.userAgentValue) => Android
                        case _                                       => PC
                    )
      os         <- ZIO.succeed(
                      command.userAgent match
                        case a if a.contains(Windows.userAgentValue)   => Windows
                        case b if b.contains(Macintosh.userAgentValue) => Macintosh
                        case c if c.contains(Linux.userAgentValue)     => Linux
                        case d                                         => OtherOS
                    )
      userAccess <- ZIO.succeed(
                      ShortenedUrlUserAccess(
                        browser = browser,
                        device = device,
                        os = os,
                        shortenedUrl = command.shortenedUrl,
                        originalUrl = command.originalUrl
                      )
                    )
                    _ <- ZIO.log("here")
      _          <- shortenedUrlUserAccessRepository.saveShortenedUrlUserAccess(
                      userAccess
                    )
    } yield ()
}

object ShortenedUrlUserEventsService {
  val layer: ZLayer[ShortenedUrlUserAccessRepository, Throwable, ShortenedUrlUserEventsService] = ZLayer.fromZIO(
    for {
      repo <- ZIO.service[ShortenedUrlUserAccessRepository]
    } yield (ShortenedUrlUserEventsServiceImpl(repo))
  )
}

File: myurlshortner-consumer/src/main/kotlin/com/acme/myurlshortner/consumer/application/service/ShortenedUrlUserEventsServiceImpl.kt
package com.acme.myurlshortner.consumer.application.service

import com.acme.myurlshortner.consumer.domain.useragent.Browser
import com.acme.myurlshortner.consumer.domain.useragent.Browser.*
import com.acme.myurlshortner.consumer.domain.useragent.Device
import com.acme.myurlshortner.consumer.domain.useragent.Device.*
import com.acme.myurlshortner.consumer.domain.useragent.OperatingSystem
import com.acme.myurlshortner.consumer.domain.useragent.OperatingSystem.*
import com.acme.myurlshortner.consumer.domain.userevent.command.UserAccessedShortenedUrlCommand
import com.acme.myurlshortner.consumer.domain.userevent.entity.UserAccessedShortenedUrl
import com.acme.myurlshortner.consumer.domain.userevent.repo.UserAccessedShortenedUrlRepo
import com.acme.myurlshortner.consumer.domain.userevent.service.ShortenedUrlUserEventsService
import org.jboss.logging.Logger
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import kotlin.random.Random

@Service
class ShortenedUrlUserEventsServiceImpl(
    private val repo: UserAccessedShortenedUrlRepo
) : ShortenedUrlUserEventsService {

    private val logger = Logger.getLogger(ShortenedUrlUserEventsServiceImpl::class.java)

    @Transactional
    override fun handleShortenedUrlUserAccessed(
        command: UserAccessedShortenedUrlCommand,
    ) {
        val randomError = Random.nextInt(1, 101)
        if (randomError in 1..50) {
            logger.info("Generated error number $randomError")
            throw IllegalArgumentException("Random error 1 <= $randomError <= 50")
        } else {
            logger.info("No Error")
        }
        repo.saveUserAccessedShortenedUrl(
            UserAccessedShortenedUrl(
                uniqueIdentifier = command.uniqueIdentifier,
                originalUrl = command.originalUrl,
                shortenedUrl = command.shortenedUrl,
                device = when {
                    command.userAgent.contains(iPhone.userAgentValue!!) -> iPhone
                    command.userAgent.contains(iPad.userAgentValue!!) -> iPad
                    command.userAgent.contains(Android.userAgentValue!!) -> Android
                    command.userAgent.contains(PC.userAgentValue!!) -> PC
                    command.userAgent.contains(Mac.userAgentValue!!) -> Mac
                    else -> Device.Other
                },
                browser = when {
                    command.userAgent.contains(Safari.userAgentValue!!) -> Safari
                    command.userAgent.contains(Chrome.userAgentValue!!) -> Chrome
                    command.userAgent.contains(Firefox.userAgentValue!!) -> Firefox
                    else -> Browser.Other
                },
                operatingSystem = when {
                    command.userAgent.contains(Windows.userAgentValue!!) -> Windows
                    command.userAgent.contains(Macintosh.userAgentValue!!) -> Macintosh
                    command.userAgent.contains(Linux.userAgentValue!!) -> Linux
                    else -> OperatingSystem.Other
                },
                accessedAt = command.accessedAt
            )
        )
    }
}
